### PR TITLE
[WIP] Kademlia

### PR DESCRIPTION
**Pull request by oskarth**
Start of Kademlia implementation, very hacky and needs to be cleaned up. Basic find node functionality with protobuf encoding/decoding so far. Still need to hook up handler, implement iterative lookup, add more more RPC calls, clean up message types, etc, etc. Opening this draft PR to share current state and to get feedback on direction.

It largely follows existing structure of pubsub. For spec, it is currently using (but nowhere near complete with): http://xlattice.sourceforge.net/components/protocol/kademlia/specs.html and somewhat inspired by kad-dht in libpp2p.

### CODE REVIEW

**Comment by arnetheduck** [comment]
since this is in nim-libp2p and is called kademlia, it should reasonably implement and interop with upstream kad-dht as a first step - or we should move it to some place that clearly points out that this is an extension that doesn't interop. 

there's another kademlia [here for discv4](https://github.com/status-im/nim-eth/blob/master/eth/p2p/kademlia.nim) and another [here for discv5](https://github.com/status-im/nim-eth/blob/master/eth/p2p/discoveryv5/routing_table.nim) for inspiration.

**Comment by oskarth** [comment]
@arnetheduck Ack and agree. FWIW, the current spec for kad-dht is currently in WIP PR for 1y+ (https://github.com/libp2p/specs/pull/108), and there's a protobuf in rust-libp2p here: https://github.com/libp2p/rust-libp2p/blob/master/protocols/kad/src/dht.proto - I did base it on this one as well but took some shortcuts (enums/replies) to get basics end to end. Will work towards compatibility with that spec, or alternatively make clear this how this is an extension.

As for other implementations, I deliberately didn't look at them yet. Now I might have a look though, thanks!

File: libp2p/protocols/kademlia/rpc/messages.nim [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  RPCMsg* = object
+    messageType*: MessageType
+    key*: Key
+    closerPeers*: Option[CloserPeers]
+
+    # Returns values
+    #record*: Option[Record]
+
+    # Return peers closer a key in a query
+    # E.g. FIND_NODE
```
Comment (zah) [inline libp2p/protocols/kademlia/rpc/messages.nim]: Accidentally committed binary file:
libp2p/protocols/kademlia/rpc/protobuf

File: libp2p/protocols/kademlia/kademlia.nim [hunk 2]
```diff
@@ -0,0 +1,521 @@
...
+  # start kad
+  discard
+
+method stop*(p: KadProto) {.async, base.} =
+  # stop kad
+  discard
+
+# When we add contact we should also getPeer or set their peer info
+# XXX: Do we want to listen to them as well?
+method addContact*(p: KadProto, contact: PeerInfo) {.base, gcsafe.} =
```
Comment (zah) [inline libp2p/protocols/kademlia/kademlia.nim]: Is the intention here for these Kademlia-specific operations to also be defined as methods? As far as I can tell, you don't intend to introduce types derived from `KadProto` and these could be regular procs.

**Comment by arnetheduck** [comment]
what's the plan here?

**Comment by oskarth** [comment]
The latest is that we want to figure out a better discovery interface for the parts that modify libp2p directly (switch, etc), then split Kademlia up into a separate repo. Waiting for @dryajov to have bandwidth for this (we talked about this at Vac PM calls a few times).

Haven't looked at this for quite a while once I'm done with GossipSub PoC I can revisit this and try to figure out how to split it up.

**Pull request closed without merging**